File: website/download-catalog.py
#!/usr/bin/env python3
"""Fetch the latest private export during CI. No source URL/credentials enter the site."""
import argparse
import base64
import hashlib
import json
import os
from pathlib import Path, PurePosixPath
import re
import tarfile
import tempfile
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit
from urllib.request import Request, build_opener, HTTPRedirectHandler

MAX_BYTES = 2 * 1024**3
METADATA = {"search3.json", "assets.json", "credits.json", "credits.html", "dates.json", "sources.json", "summary.json"}
FOLDERS = {"audio", "covers", "albums", "artists", "evidence"}


class ExportError(Exception):
    """Safe to log: messages must never contain source URLs or credentials."""


class NoRedirects(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        raise ExportError("Export source redirected; refusing to forward credentials.")


def source_url(value):
    try:
        parsed = urlsplit(value)
        valid = parsed.scheme == "https" and parsed.hostname and not (parsed.username or parsed.password or parsed.query or parsed.fragment)
        if not valid:
            raise ValueError()
        return value.rstrip("/") + "/"
    except ValueError:
        raise ExportError("DEMO_EXPORT_URL must be an HTTPS export-directory URL without credentials, query or fragment.") from None


def file_inventory(manifest):
    files = manifest["files"]
    if not isinstance(files, list) or len(files) > 10000:
        raise ExportError("Invalid export inventory.")
    indexed = {}
    total = 0
    for entry in files:
        name, size, checksum = entry["path"], entry["size"], entry["sha256"]
        path = PurePosixPath(name)
        if (not name or path.is_absolute() or ".." in path.parts or "\\" in name
                or path.as_posix() != name or name in indexed
                or (path.parts[0] not in FOLDERS and name not in METADATA)
                or not isinstance(size, int) or size < 0
                or not re.fullmatch(r"[a-f0-9]{64}", checksum)):
            raise ExportError("Unsafe or invalid export inventory.")
        total += size
        indexed[name] = entry
    if total > MAX_BYTES or not METADATA.issubset(indexed):
        raise ExportError("Incomplete or oversized export inventory.")
    return indexed


def extract(archive, output, files, forbidden):
    seen = set()
    with tarfile.open(archive, "r:gz") as tar:
        for member in tar:
            entry = files.get(member.name)
            if not member.isfile() or not entry or member.name in seen or member.size != entry["size"]:
                raise ExportError("Unexpected archive entry.")
            target = output / member.name
            target.parent.mkdir(parents=True, exist_ok=True)
            digest = hashlib.sha256()
            overlap = b""
            longest = max(map(len, forbidden), default=1)
            with tar.extractfile(member) as source, target.open("xb") as destination:
                for block in iter(lambda: source.read(1024 * 1024), b""):
                    check = overlap + block
                    if any(secret in check for secret in forbidden):
                        raise ExportError("Export contains a private source identifier or credential; refusing publication.")
                    overlap = check[-longest:]
                    digest.update(block)
                    destination.write(block)
            if digest.hexdigest() != entry["sha256"]:
                raise ExportError("Extracted file checksum mismatch.")
            seen.add(member.name)
    if seen != files.keys():
        raise ExportError("Archive is incomplete.")


def download(output, export_url, password, username="demo", opener=None):
    if output.exists():
        raise ExportError("Catalog destination already exists; refusing to mix releases.")
    if not export_url or not password or not username:
        raise ExportError("Set DEMO_EXPORT_URL and DEMO_MEDIA_PASSWORD in Actions secrets.")
    base = source_url(export_url)
    authorization = "Basic " + base64.b64encode((username + ":" + password).encode()).decode()
    opener = opener or build_opener(NoRedirects())

    def get(path):
        try:
            return opener.open(Request(base + path, headers={"Authorization": authorization}), timeout=120)
        except HTTPError as error:
            raise ExportError(f"Export download failed (HTTP {error.code}); check the source secrets.") from None
        except (URLError, OSError):
            raise ExportError("Could not reach the export source.") from None

    def metadata(path):
        with get(path) as response:
            raw = response.read(8 * 1024 * 1024 + 1)
            if len(raw) > 8 * 1024 * 1024:
                raise ExportError("Export metadata exceeds size limit.")
            return json.loads(raw)

    # Resolve latest once, then use only the immutable release for this build.
    release = metadata("latest.json")["release"]
    if not isinstance(release, str) or not re.fullmatch(r"[a-f0-9]{64}", release):
        raise ExportError("Invalid export release.")
    manifest = metadata(release + "/manifest.json")
    fingerprint = hashlib.sha256(json.dumps(manifest["files"], sort_keys=True, separators=(",", ":")).encode()).hexdigest()
    if manifest["schemaVersion"] != 1 or manifest["release"] != release or fingerprint != release:
        raise ExportError("Invalid export manifest fingerprint.")
    files = file_inventory(manifest)
    item = manifest["archives"]["full"]
    if (item["path"] != "catalog.tar.gz" or not isinstance(item["size"], int)
            or not 0 < item["size"] <= MAX_BYTES
            or not re.fullmatch(r"[a-f0-9]{64}", item["sha256"])):
        raise ExportError("Invalid archive metadata.")
    output.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.TemporaryDirectory(prefix=".demo-download-", dir=output.parent) as tmp:
        root = Path(tmp)
        archive = root / "catalog.tar.gz"
        digest, size = hashlib.sha256(), 0
        with get(release + "/catalog.tar.gz") as response, archive.open("wb") as destination:
            for block in iter(lambda: response.read(1024 * 1024), b""):
                size += len(block)
                if size > item["size"]:
                    raise ExportError("Archive exceeds declared size.")
                digest.update(block)
                destination.write(block)
        if size != item["size"] or digest.hexdigest() != item["sha256"]:
            raise ExportError("Archive checksum mismatch.")
        stage = root / "catalog"
        stage.mkdir()
        # Check the raw export before copying it to Pages. Check across chunk boundaries too.
        forbidden = [base.encode(), urlsplit(base).hostname.encode(), password.encode(), authorization.encode()]
        extract(archive, stage, files, forbidden)
        search = json.loads((stage / "search3.json").read_text())["subsonic-response"]
        if search["status"] != "ok" or not search["searchResult3"]["song"]:
            raise ExportError("Export has no usable music catalog.")
        assets = json.loads((stage / "assets.json").read_text())
        for asset in assets.values():
            entry = files.get(asset["path"])
            if not entry or (asset.get("sha256") and entry["sha256"] != asset["sha256"]):
                raise ExportError("Export references a missing or corrupt asset.")
        stage.rename(output)
    return release


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    try:
        release = download(args.output, os.environ.get("DEMO_EXPORT_URL", ""),
                           os.environ.get("DEMO_MEDIA_PASSWORD", ""), os.environ.get("DEMO_MEDIA_USERNAME", "demo"))
    except ExportError as error:
        parser.exit(1, f"Demo export failed: {error}\n")
    except Exception:
        # Library exceptions can embed request URLs; never print their repr/traceback in CI.
        parser.exit(1, "Demo export failed: invalid or unavailable catalog.\n")
    print("Verified demo export:", release)


if __name__ == "__main__":
    main()

File: website/download-catalog.mjs
import { createHash } from "node:crypto";
import { createReadStream } from "node:fs";
import { chmod, lstat, mkdir, mkdtemp, rename, rm } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";
import { parseArgs } from "node:util";
import { createGunzip } from "node:zlib";
import { unpackTar } from "modern-tar/fs";

const MAX_BYTES = 2 * 1024 ** 3;
const METADATA = new Set([
  "search3.json",
  "assets.json",
  "credits.json",
  "credits.html",
  "dates.json",
  "sources.json",
  "summary.json",
]);
const FOLDERS = new Set(["audio", "covers", "albums", "artists", "evidence"]);
const checksum = (value) => typeof value === "string" && /^[a-f0-9]{64}$/.test(value);
const hash = (value) => createHash("sha256").update(value).digest("hex");
class ExportError extends Error {}
function requireValid(condition, message) {
  if (!condition) throw new ExportError(message);
}

// Bound streams, verify bytes and scan across chunk boundaries without buffering audio files.
function verifyBytes(limit, expectedHash, forbidden = []) {
  let size = 0;
  const digest = createHash("sha256");
  const overlapSize = Math.max(0, ...forbidden.map((value) => value.length - 1));
  let overlap = Buffer.alloc(0);
  return new Transform({
    transform(chunk, _encoding, callback) {
      size += chunk.length;
      if (size > limit) return callback(new ExportError("Export exceeds declared size."));
      const scanned = Buffer.concat([overlap, chunk]);
      if (forbidden.some((value) => scanned.includes(value))) {
        return callback(
          new ExportError("Export contains a private source identifier or credential."),
        );
      }
      overlap = overlapSize ? Buffer.from(scanned.subarray(-overlapSize)) : Buffer.alloc(0);
      digest.update(chunk);
      callback(null, chunk);
    },
    flush(callback) {
      if (expectedHash && (size !== limit || digest.digest("hex") !== expectedHash)) {
        return callback(new ExportError("Export checksum or size mismatch."));
      }
      callback();
    },
  });
}

async function download(output) {
  const { DEMO_EXPORT_URL, DEMO_MEDIA_PASSWORD, DEMO_MEDIA_USERNAME = "demo" } = process.env;
  requireValid(
    DEMO_EXPORT_URL && DEMO_MEDIA_PASSWORD && DEMO_MEDIA_USERNAME,
    "Set DEMO_EXPORT_URL and DEMO_MEDIA_PASSWORD in Actions secrets.",
  );
  let base;
  try {
    base = new URL(DEMO_EXPORT_URL);
  } catch {
    throw new ExportError("Invalid DEMO_EXPORT_URL.");
  }
  requireValid(
    base.protocol === "https:" && !base.username && !base.password && !base.search && !base.hash,
    "DEMO_EXPORT_URL must be an HTTPS directory without credentials, query or fragment.",
  );
  if (!base.pathname.endsWith("/")) base.pathname += "/";
  const authorization =
    "Basic " + Buffer.from(`${DEMO_MEDIA_USERNAME}:${DEMO_MEDIA_PASSWORD}`).toString("base64");
  const forbidden = [base.href, base.hostname, DEMO_MEDIA_PASSWORD, authorization].map((value) =>
    Buffer.from(value),
  );
  const existing = await lstat(output).catch((error) => {
    if (error.code !== "ENOENT") throw error;
  });
  requireValid(!existing, "Catalog destination already exists; refusing to mix releases.");

  async function get(path) {
    const response = await fetch(new URL(path, base), {
      headers: { Authorization: authorization },
      redirect: "error",
      signal: AbortSignal.timeout(10 * 60 * 1000),
    });
    if (!response.ok) {
      await response.body?.cancel();
      throw new ExportError(
        `Export download failed (HTTP ${response.status}); check the source secrets.`,
      );
    }
    return response.body;
  }
  async function metadata(path) {
    const chunks = [];
    let size = 0;
    for await (const chunk of await get(path)) {
      size += chunk.length;
      requireValid(size <= 8 * 1024 ** 2, "Export metadata exceeds size limit.");
      chunks.push(chunk);
    }
    return JSON.parse(Buffer.concat(chunks).toString("utf8"));
  }

  // Resolve latest once; every subsequent request uses that immutable release.
  const { release } = await metadata("latest.json");
  requireValid(checksum(release), "Invalid export release.");
  const manifest = await metadata(`${release}/manifest.json`);
  requireValid(
    Array.isArray(manifest.files) && manifest.files.length <= 10000,
    "Invalid export inventory.",
  );
  // Match the exporter's sorted-key, ASCII-escaped canonical JSON fingerprint.
  const canonical = JSON.stringify(
    manifest.files.map((entry) =>
      Object.fromEntries(Object.entries(entry).sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))),
    ),
  ).replace(/[\u007f-\uffff]/g, (char) => "\\u" + char.charCodeAt(0).toString(16).padStart(4, "0"));
  requireValid(
    manifest.schemaVersion === 1 && manifest.release === release && hash(canonical) === release,
    "Invalid export manifest fingerprint.",
  );
  const files = new Map();
  let total = 0;
  for (const entry of manifest.files) {
    const { path, size, sha256 } = entry;
    requireValid(typeof path === "string" && path.length > 0, "Invalid export path.");
    const parts = path.split("/");
    requireValid(
      !/[\\\x00-\x1f]/.test(path) &&
        parts.every((part) => part && part !== "." && part !== "..") &&
        (FOLDERS.has(parts[0]) || METADATA.has(path)) &&
        !files.has(path) &&
        Number.isSafeInteger(size) &&
        size >= 0 &&
        checksum(sha256) &&
        !forbidden.some((value) => Buffer.from(path).includes(value)),
      "Unsafe or invalid export inventory.",
    );
    total += size;
    files.set(path, entry);
  }
  requireValid(
    total <= MAX_BYTES && [...METADATA].every((name) => files.has(name)),
    "Incomplete or oversized export inventory.",
  );
  const item = manifest.archives.full;
  requireValid(
    item.path === "catalog.tar.gz" &&
      Number.isSafeInteger(item.size) &&
      item.size > 0 &&
      item.size <= MAX_BYTES &&
      checksum(item.sha256),
    "Invalid archive metadata.",
  );

  await mkdir(dirname(output), { recursive: true });
  const temporary = await mkdtemp(join(dirname(output), ".demo-download-"));
  try {
    const seen = new Set();
    await pipeline(
      await get(`${release}/catalog.tar.gz`),
      verifyBytes(item.size, item.sha256),
      createGunzip(),
      verifyBytes(MAX_BYTES + 64 * 1024 ** 2),
      unpackTar(temporary, {
        strict: true,
        fmode: 0o644,
        dmode: 0o755,
        filter({ name, size, type }) {
          const entry = files.get(name);
          requireValid(
            type === "file" && entry && !seen.has(name) && size === entry.size,
            "Unexpected archive entry.",
          );
          seen.add(name);
          return true;
        },
      }),
    );
    requireValid(seen.size === files.size, "Archive is incomplete.");
    for (const { path, size, sha256 } of files.values()) {
      await pipeline(
        createReadStream(join(temporary, path)),
        verifyBytes(size, sha256, forbidden),
        new WritableStream(),
      );
    }
    await chmod(temporary, 0o755);
    await rename(temporary, output);
    console.log("Verified demo export:", release);
  } finally {
    await rm(temporary, { recursive: true, force: true });
  }
}

try {
  const { values } = parseArgs({ options: { output: { type: "string" } } });
  requireValid(values.output, "Provide --output <catalog directory>.");
  await download(resolve(values.output));
} catch (error) {
  // Fetch/filesystem/library exceptions can contain private URLs. Never print them.
  console.error(
    "Demo export failed:",
    error instanceof ExportError ? error.message : "Invalid or unavailable catalog.",
  );
  process.exitCode = 1;
}
